Order Creation: Sync product list from remote in Add Product screen

rachelmcr opened the pull request:
Part of: #5408

## Description

In Order Creation, the Add Product screen now syncs the product list from remote, with the following behavior:

- When the Add Product screen first opens, and no products exist in storage, a ghost list view appears. Once the first page of products are synced, the ghost list is replaced by a real list of products.
- When the Add Product screen first opens, and there are products in storage, the stored products appear immediately (with no ghost view).
- If there are more products to sync from remote, an infinite scroll indicator appears at the bottom of the list while the next page of products are syncing.

Several notes/caveats from #5589 still apply:

- For now we're only loading non-variable products. (Variable products will be implemented later.)
- The product list only shows published and private products. This matches core/web behavior, where draft and pending products can't be added to an order.
- For now we're only showing a placeholder product image. (Product images will be added in another PR.)

## Changes

* `AddProductToOrder`:
   * The `AddProductToOrder` view now switches on `SyncStatus` to display the appropriate view: a list of products if there are any products to show, an empty state if there are no products, a ghost list during the initial sync, and an empty view as a default (initial state).
   * When `AddProductToOrder` appears, the initial sync is triggered. Subsequent syncs are triggered when the last row of the product list appears.
* `AddProductToOrderViewModel`:
   * The list of products is now filtered in `products` instead of the results controller. This allows us to use the full list of products in the results controller to trigger the next page for syncing.
   * Adds a `SyncCoordinator` and necessary methods to sync products from remote, update the `syncStatus` before/during/after sync, and update the results controller after sync.
* `ProductRowViewModel` is now `Equatable` to be able to check when a product row is the last row in the list.
* Adds unit tests for the scrolling indicator and sync status changes.

## Testing

1. Build and run the app in debug mode (to ensure the feature flag is enabled).
2. Make sure Order Creation is enabled under Settings > Experimental Features.
3. Go to the Orders tab and tap the + button. (If Simple Payments is also enabled, tap "Create order" in the bottom sheet that appears.)
4. On the New Order screen, tap the "Add product" button.
5. On the Add Product screen, confirm you see the expected behavior based on whether you have any products in storage and whether your store has any non-variable, published or private products to sync from remote.

Scenarios to consider:

* If you have not synced any products from remote yet (e.g. log in and immediately create a new order, without opening the Products tab) you should see the ghost list while the first sync is happening.
* If you have synced _all_ of your store's products from remote, the full list of non-variable, published or private products should be loaded in the product list.
* If you have synced _some_ but not all of your store's products from remote, when you scroll down you should see the scroll indicator (spinner) at the bottom of the list while more products are being synced.
* If you don't have any products on your store, you should see an empty view.

## Screenshots


https://user-images.githubusercontent.com/8658164/145828789-2d711050-c08b-445d-9741-62dbdd5bc2eb.mp4



Empty view|Loading view (first sync)|Product list
-|-|-
![Simulator Screen Shot - iPhone 13 Pro - 2021-12-13 at 12 37 19](https://user-images.githubusercontent.com/8658164/145828699-934864d6-582e-4df8-9973-aaf62d7fc77e.png)|![Simulator Screen Shot - iPhone 13 Pro - 2021-12-13 at 12 37 58](https://user-images.githubusercontent.com/8658164/145828709-8c5f172d-d1a3-4bc5-8786-84b029800f8e.png)|![Simulator Screen Shot - iPhone 13 Pro - 2021-12-13 at 12 38 01](https://user-images.githubusercontent.com/8658164/145828718-c13648ad-2b2b-4e5f-bb6e-ad74e42d78de.png)


## Submitter Checklist

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

rachelmcr committed:
Sync products from remote if needed

rachelmcr committed:
Try syncing more products when the last product in the list appears

rachelmcr committed:
Fix list style for ghost rows

rachelmcr committed:
Make sure infinite scroll indicator appears at end of list

rachelmcr committed:
Clarify sync actions and state

rachelmcr committed:
Return all products from storage and filter product list separately

rachelmcr committed:
Use syncing coordinator to differentiate between synced results controller and filtered product list

rachelmcr committed:
Display scrolling indicator while in syncing state

rachelmcr committed:
Remove accidentally re-added view

rachelmcr committed:
Add sync unit tests

rachelmcr committed:
Update comments

rachelmcr commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/ProductsSection/AddProductToOrder.swift:
```diff
@@ -14,15 +14,39 @@ struct AddProductToOrder: View {
-                .frame(maxHeight: .infinity)
-                .renderedIf(viewModel.productRows.isEmpty)
+                        // Infinite scroll indicator
+                        ProgressView()
+                            .frame(maxWidth: .infinity, alignment: .center)
+                            .listRowInsets(EdgeInsets())
+                            .listRowBackground(Color(.listBackground))
+                            .renderedIf(viewModel.shouldShowScrollIndicator)
```
This view is loaded as the last item in the list, and unfortunately I can't find a way to remove the separator below the row except with `.listRowSeparator(.hidden)` which is iOS 15+.

Is it worth using an `if #available` check to add that modifier, and otherwise just show this (for devices with iOS <15)?

ealeksandrov replied:
I think yes, if it doesn't require much effort.

rachelmcr replied:
Yep, this turned out to be a simple check, added in a325ea6.

ealeksandrov commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/ProductsSection/AddProductToOrderViewModel.swift:
```diff
@@ -29,31 +40,155 @@ final class AddProductToOrderViewModel: ObservableObject {
+            case .failure(let error):
+                DDLogError("⛔️ Error synchronizing products during order creation: \(error)")
+            case .success:
+                self.updateProductsResultsController()
```
nit: `success` should come before `failure`

rachelmcr replied:
Updated in 1e5d0d9.

ealeksandrov commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/ProductsSection/AddProductToOrderViewModel.swift:
```diff
@@ -29,31 +40,155 @@ final class AddProductToOrderViewModel: ObservableObject {
+        case firstPageSync
+        case results
+        case empty
+        case none
```
nit: `none` is used only as initial value. Do we really need it?

rachelmcr replied:
Good point! Removed that case in 0a0dece.

ealeksandrov reviewed (approved):
Works well in all states 🚀
Code is good and well commented - spotted just a few minor nits.

One visual nit I've noticed - there is no safe area margin in the bottom of a list, so last product cell is hidden behind home indicator:
<img width=350 src=https://user-images.githubusercontent.com/3132438/146007376-e82e3554-68d4-45bf-bdd3-379e51f19868.png>

rachelmcr committed:
Put success case before failure case

rachelmcr committed:
Remove unneeded initial value from SyncStatus

rachelmcr committed:
Hide list row separator below scroll indicator on iOS 15

rachelmcr commented:
Thanks for the review @ealeksandrov! I addressed all the feedback except the visual issue you mentioned:

> One visual nit I've noticed - there is no safe area margin in the bottom of a list, so last product cell is hidden behind home indicator

For some reason I can't reproduce this on my end right now — I always see a bottom margin when I reach the end of the list — but I've opened https://github.com/woocommerce/woocommerce-ios/issues/5677 to take a closer look at it separately. (This also came up as an issue on the New Order screen in https://github.com/woocommerce/woocommerce-ios/pull/5675#pullrequestreview-831530272.)
